src/pages/studyView/StudyViewUtils.tsx

```diff
@@ -3022,6 +2971,87 @@ export function updateSavedUserPreferenceChartIds(
     return chartSettings;
 }
 
+export async function getAllClinicalDataByStudyViewFilter(
+    studyViewFilter: StudyViewFilter
+): Promise<{ [sampleId: string]: { [attributeId: string]: string } }> {
+    const localClinicalDataCollection: ClinicalDataCollection = {
+        sampleClinicalData: [],
+        patientClinicalData: [],
+    };
+    let remoteClinicalDataCollection: ClinicalDataCollection = {
+        sampleClinicalData: [],
+        patientClinicalData: [],
+    };
+
+    const maxPageSize = 500000;
+    let pageNumber = 0;
+
+    do {
+        const remoteClinicalDataCollection = await internalClient.fetchClinicalDataClinicalTableUsingPOST(
+            {
+                studyViewFilter,
+                pageSize: maxPageSize,
+                pageNumber: pageNumber,
+                searchTerm: undefined,
+                sortBy: undefined,
+                direction: 'ASC',
+            }
+        );
+        localClinicalDataCollection.sampleClinicalData = localClinicalDataCollection.sampleClinicalData.concat(
+            remoteClinicalDataCollection.sampleClinicalData
+        );
+        localClinicalDataCollection.patientClinicalData = localClinicalDataCollection.patientClinicalData.concat(
+            remoteClinicalDataCollection.patientClinicalData
+        );
+        pageNumber++;
+    } while (remoteClinicalDataCollection.sampleClinicalData.length > 0);
+
+    return mergeClinicalDataCollection(localClinicalDataCollection);
+}
+
+export function mergeClinicalDataCollection(
+    clinicalDataCollection: ClinicalDataCollection
+): { [sampleId: string]: { [attributeId: string]: string } } {
+    const patientKeyedSampleData = _.groupBy(
+        clinicalDataCollection.sampleClinicalData,
+        d => d.uniquePatientKey
+    );
+    let patientKeyedSampleKeyedData = _.mapValues(
+        patientKeyedSampleData,
+        sampleData => _.groupBy(sampleData, d => d.uniqueSampleKey)
+    );
+    const patientKeyedPatientData = _.groupBy(
+        clinicalDataCollection.patientClinicalData,
+        d => d.uniquePatientKey
+    );
+    // Add patient level clinical data to sample clinical data.
+    patientKeyedSampleKeyedData = _.mapValues(
+        patientKeyedSampleKeyedData,
+        (sampleKeyedData, patientId) =>
+            _.mapValues(sampleKeyedData, (attrs, sampleId) =>
+                attrs.concat(patientKeyedPatientData[patientId] || [])
+            )
+    );
+    // Remove patient id levels (only keep sample id keys).
+    const sampleKeyedData = _.assign(
```

Comment:
this seems odd to me.  you are assigning an array (_.values) to an object, which seems to produce an object with numerical indexes as keys { 0:thing, 1:thing, 2:thing  }.  That seems analagous to an array and should probablu just be an array?

Reply:
No, `sampleKeyedData` are arrays of clinical data keyed by unique sample id like so:

```
{
  s1key => [clinical_atrr1, clinical_attr2, ...]
}
```

In `sampleCollapsedAttributes` this is collapsed into the format accepted by the Clinical Table like so:

```
{
  s1key => {
      clinical_atrr1: clinical_value1,
      clinical_attr2: clinicel_value2
  }
}
```

So, the keying in `sampleKeyedData` is needed to reference all clinical data of a sample. An array is not suitable in this case.

Reply:
i see, so values produces an array of objects which have sampleIds as keys and we are just copying them onto one object